### PR TITLE
[Turbopack] replace EnvLayer with a faster filter

### PR DESCRIPTION
**Pull request by sokra**
### What?

The default `EnvFilter` is very flexible, but also very slow.

This replaces it with a custom one that is a lot faster.

Closes PACK-3624

### CODE REVIEW

**Comment by sokra** [comment]
* **#73615** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73615?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#73614** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73614?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/vercel/next.js/73614?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `canary`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 00d22741f4a087dca11bf27a6dd4f65263f1ebf1

`pnpm test-dev test/development/acceptance-app/ReactRefreshLogBox.test.ts`
- ReactRefreshLogBox app default > server component can recover from error thrown in the module

<details>
<summary>Expand output</summary>

  ● ReactRefreshLogBox app default › server component can recover from error thrown in the module

    Expected Redbox but found none

      1060 |
      1061 |       await next.patchFile('index.js', "throw new Error('module error')")
    > 1062 |       await session.assertHasRedbox()
           |       ^
      1063 |       await next.patchFile(
      1064 |         'index.js',
      1065 |         'export default function Page() {return <p>hello world</p>}'

      at Object.<anonymous> (development/acceptance-app/ReactRefreshLogBox.test.ts:1062:7)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| buildDuration | 20.8s | 17.9s | N/A |
| buildDurationCached | 17s | 14.9s | N/A |
| nodeModulesSize | 409 MB | 409 MB | N/A |
| nextStartRea..uration (ms) | 459ms | 456ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| 1187-HASH.js gzip | 50.1 kB | 50.1 kB | N/A |
| 8276.HASH.js gzip | 169 B | 168 B | N/A |
| 8377-HASH.js gzip | 5.3 kB | 5.3 kB | N/A |
| bccd1874-HASH.js gzip | 53 kB | 53 kB | N/A |
| framework-HASH.js gzip | 57.5 kB | 57.5 kB | N/A |
| main-app-HASH.js gzip | 233 B | 235 B | N/A |
| main-HASH.js gzip | 33.7 kB | 33.7 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 39.4 kB | 39.4 kB | ✓ |
| Overall change | 39.4 kB | 39.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 193 B | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 513 B | 511 B | N/A |
| css-HASH.js gzip | 343 B | 342 B | N/A |
| dynamic-HASH.js gzip | 1.84 kB | 1.84 kB | N/A |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 363 B | 362 B | N/A |
| hooks-HASH.js gzip | 393 B | 392 B | N/A |
| image-HASH.js gzip | 4.44 kB | 4.43 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.35 kB | 2.34 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 328 B | ✓ |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 323 B | 326 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.75 kB | 1.75 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| _buildManifest.js gzip | 746 B | 747 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| index.html gzip | 523 B | 524 B | N/A |
| link.html gzip | 537 B | 537 B | ✓ |
| withRouter.html gzip | 519 B | 521 B | N/A |
| Overall change | 537 B | 537 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| edge-ssr.js gzip | 128 kB | 128 kB | N/A |
| page.js gzip | 202 kB | 202 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 666 B | 665 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 31.1 kB | 31.1 kB | N/A |
| edge-runtime..pack.js gzip | 844 B | 844 B | ✓ |
| Overall change | 844 B | 844 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| 797-experime...dev.js gzip | 322 B | 322 B | ✓ |
| 797.runtime.dev.js gzip | 314 B | 314 B | ✓ |
| app-page-exp...dev.js gzip | 322 kB | 322 kB | ✓ |
| app-page-exp..prod.js gzip | 126 kB | 126 kB | ✓ |
| app-page-tur..prod.js gzip | 139 kB | 139 kB | ✓ |
| app-page-tur..prod.js gzip | 135 kB | 135 kB | ✓ |
| app-page.run...dev.js gzip | 312 kB | 312 kB | ✓ |
| app-page.run..prod.js gzip | 122 kB | 122 kB | ✓ |
| app-route-ex...dev.js gzip | 36.8 kB | 36.8 kB | ✓ |
| app-route-ex..prod.js gzip | 25 kB | 25 kB | ✓ |
| app-route-tu..prod.js gzip | 25 kB | 25 kB | ✓ |
| app-route-tu..prod.js gzip | 24.8 kB | 24.8 kB | ✓ |
| app-route.ru...dev.js gzip | 38.5 kB | 38.5 kB | ✓ |
| app-route.ru..prod.js gzip | 24.8 kB | 24.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.56 kB | 9.56 kB | ✓ |
| pages-api.ru...dev.js gzip | 11.4 kB | 11.4 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.56 kB | 9.56 kB | ✓ |
| pages-turbo...prod.js gzip | 21.3 kB | 21.3 kB | ✓ |
| pages.runtim...dev.js gzip | 27 kB | 27 kB | ✓ |
| pages.runtim..prod.js gzip | 21.3 kB | 21.3 kB | ✓ |
| server.runti..prod.js gzip | 916 kB | 916 kB | ✓ |
| Overall change | 2.35 MB | 2.35 MB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js sokra/perf-tracing | Change |
| - | - | - | - |
| 0.pack gzip | 2.03 MB | 2.03 MB | ⚠️ +776 B |
| index.pack gzip | 145 kB | 146 kB | ⚠️ +142 B |
| Overall change | 2.18 MB | 2.18 MB | ⚠️ +918 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 00d22741f4a087dca11bf27a6dd4f65263f1ebf1

File: turbopack/crates/turbopack-trace-utils/src/filter_layer.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        metadata: &tracing::Metadata<'_>,
+        _ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        if self.config.is_empty() {
+            return true;
+        }
+        let target = metadata.target().split("::").next().unwrap();
+        let filter = self.config.get(target).unwrap_or(&self.global_level);
+        let level = metadata.level();
+        return level <= filter;
```
Comment (mischnic) [inline turbopack/crates/turbopack-trace-utils/src/filter_layer.rs:50]:
```suggestion
       level <= filter
```

File: turbopack/crates/turbopack-trace-utils/src/filter_layer.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        metadata: &tracing::Metadata<'_>,
+        _ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        if self.config.is_empty() {
+            return true;
+        }
+        let target = metadata.target().split("::").next().unwrap();
+        let filter = self.config.get(target).unwrap_or(&self.global_level);
+        let level = metadata.level();
+       level <= filter
```
Comment (mischnic) [inline turbopack/crates/turbopack-trace-utils/src/filter_layer.rs:50]:
```suggestion
        level <= filter
```

**Review by mischnic — APPROVED** [review]

**Comment by sokra** [comment]
### Merge activity

* **Dec 9, 4:51 AM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/73614).

**Pull request merged**